FeedStoreChallenge/CodeDataFeedStore.swift

```
@@ -0,0 +1,101 @@
+//
+//  CodeDataFeedStore.swift
+//  FeedStoreChallenge
+//
+//  Created by Carlos Linares on 10/11/20.
+//  Copyright © 2020 Essential Developer. All rights reserved.
+//
+
+import CoreData
+
+public enum StoreType {
+    case persistent, inMemory
+}
+
+public class CoreDataFeedStore: FeedStore {
+        
+    enum CoreDataFeedStore: Error {
+        case modelNotFound
+    }
+    private var persistentContainer: NSPersistentContainer
+    
+    private lazy var context: NSManagedObjectContext = { persistentContainer.newBackgroundContext() }()
+    
+    public init(storeURL: URL? = nil) throws {
+        let modelName = "FeedCacheModel"
+        let modelextension = "momd"
+        guard
+            let modelURL = Bundle(for: type(of: self)).url(forResource: modelName, withExtension: modelextension),
+            let mom = NSManagedObjectModel(contentsOf: modelURL) else {
+            throw CoreDataFeedStore.modelNotFound
+        }
+        
+        persistentContainer = NSPersistentContainer(name: modelName, managedObjectModel: mom)
+
+        if let url = storeURL {
+            let description = NSPersistentStoreDescription()
+            description.url = url
+            persistentContainer.persistentStoreDescriptions = [description]
+        }
+        persistentContainer.loadPersistentStores { _, _ in }
```

`loadPersistentStores` may fail, which would make this class unusable. But the error is ignored.

It'd be best to handle the errors (e.g., throw so clients know the instance couldn't be created correctly).

---

Thanks for reviewing my PR Caio!

I went to the implementation that you did in the course because I was lost on how can I capture the throw the error inside the closure. I thought it ran asynchronously (because is marked as _escaping_), like the init process can finish before and the closure of the `loadPersistentStores` can be called later. Can you help me understanding this?.

---

Hi @calosth - the persistent store is loaded synchronously by default (even though it's closure based!). But you can also run it asynchronously by setting the `NSPersistentStoreDescription.shouldAddStoreAsynchronously` property to `true`. 

Learn more at https://developer.apple.com/documentation/coredata/nspersistentstoredescription/1640599-shouldaddstoreasynchronously